### DISCUSSION TITLE
missing the 'build_editable' hook

### DISCUSSION THREAD

**Discussion by sam-s** [pypa/setuptools#4037] [Q&A] [answered]
`pip3 install --user -e .` fails after I switched form a trivial `setup.cfg` to an equally trivial `pyproject.toml`:
```
  Checking if build backend supports build_editable ... done
ERROR: Project file:///home/sds/src/pyapp has a 'pyproject.toml' and its build backend is missing the 'build_editable' hook. Since it does not have a 'setup.py' nor a 'setup.cfg', it cannot be installed in editable mode. Consider using a build backend that supports PEP 660.
```
despite [PEP 660 being supported by setuptools since v64](https://stackoverflow.com/q/69711606/850781) and 
```
$ pip3 install --user --upgrade setuptools
Requirement already satisfied: setuptools in /usr/lib/python3/dist-packages (59.6.0)
Collecting setuptools
  Using cached setuptools-68.1.2-py3-none-any.whl (805 kB)
Installing collected packages: setuptools
Successfully installed setuptools-68.1.2
```
and 
```
$ python3 -c 'import setuptools; print(setuptools.__version__)'
68.1.2
```

**Comment by webknjaz** [comment]
PEP 517 installs own setuptools into a throwaway tmp venv. Whatever you have in yours is never used, except for when PEP 517 isn't invoked at all.

**Comment by abravalheri** [comment, marked as answer]
Yeap, @webknjaz is right. Please make sure `pip` is updated, and that pyapp is not capping the version of setuptools specified in `pyproject.toml`.

If the problem still persists after you update pip and verify pyapp does not cap setuptools version, you can still try installing both `setuptools` and `wheel` and then running `pip install --no-build-isolation ...`. There might be a chance there is a leak on the build environment isolation provided by pip (depending on the patches your OS might apply to Python/pip).

**Reply by sam-s** [reply]
amazingly enough, 
```
$ python3 -m pip install --upgrade pip
Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: pip in /usr/lib/python3/dist-packages (22.0.2)
Collecting pip
  Downloading pip-23.2.1-py3-none-any.whl (2.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 19.7 MB/s eta 0:00:00

Installing collected packages: pip
Successfully installed pip-23.2.1
```
solved the problem
Thank you!

**Reply by abravalheri** [reply]
Yeah, that is what I suspected. Probably related to https://github.com/pypa/pip/issues/11430.

**Comment by sam-s** [comment]
> verify pyapp does not cap setuptools version

what does this mean?
`pyproject.toml` has
```
[build-system]
requires = [
    "setuptools",
...
```
I do not specify any version numbers anywhere

**Reply by abravalheri** [reply]
Yeap, that is what I meant, to look into `pyproject.toml` and check if there was no cap there.

**Reply by sam-s** [reply]
what is `cap`?

**Reply by abravalheri** [reply]
Sorry that is probably a bit of slang. It would mean a upper bound for the version.

This article also uses the `cap` slang, so I thought it was popular among devs: https://iscinumpy.dev/post/bound-version-constraints/
